### PR TITLE
Adds hbase_ycsb benchmark.

### PR DESCRIPTION
**Pull request by cmccoy**
* Adds hbase_ycsb benchmark.
* Adds a YCSB executor, result parser, tests.

Also makes starting YARN optional for Hadoop; HBase requires HDFS, but not
YARN.

By default, this runs YCSB workloads A and B against HBase, 32 threads per client VM, with an initial database size of 1GB. Each workload runs for at most 30 minutes.

I have changes ready to convert MongoDB to this framework, and add some other
databases once this PR is in.

### CODE REVIEW

**Comment by cmccoy** [comment]
(sorry this is enormous).

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+  Raises:
+    perfkitbenchmarker.data.ResourceNotFound: On missing resource.
+  """
+  hbase.CheckPrerequisites()
+  hadoop.CheckPrerequisites()
+  ycsb.CheckPrerequisites()
+
+
+def _InstallYCSB(vm):
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: any reason why InstallHbase() is nested next to runthreaded and this is defined at the top level?

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  """Install YCSB on 'vm'."""
+  vm.Install('ycsb')
+
+
+def _CreateYCSBTable(vm, table_name=TABLE_NAME, family=COLUMN_FAMILY,
+                     n_splits=TABLE_SPLIT_COUNT):
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: doc string?

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                    context={'table_name': table_name,
+                             'family': family,
+                             'n_splits': n_splits,
+                             'use_snappy': FLAGS.hbase_use_snappy})
+  # TODO(connormccoy): on HBase update, add '-n' flag.
+  command = "{0}/hbase shell {1}".format(hbase.HBASE_BIN, remote)
+  vm.RemoteCommand(command, should_log=True)
+
+
+def _GetVMsByRole(vms):
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: doc string?

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  vms = benchmark_spec.vms
+  by_role = _GetVMsByRole(benchmark_spec.vms)
+
+  loaders = by_role['loaders']
+  assert loaders, vms
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: May want to consider something other than "vms" like "Not enough loaders"

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+  # HBase cluster
+  hbase_vms = by_role['hbase_vms']
+  assert hbase_vms, vms
+  master = by_role['master']
+  zk_quorum = by_role['zk_quorum']
+  assert zk_quorum
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: Add an error message?

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  workers = by_role['workers']
+  assert workers
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: error message?

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    A list of sample.Sample objects.
+  """
+  vms = benchmark_spec.vms
+  loaders = _GetVMsByRole(vms)['loaders']
+  logging.info('Loaders: %s', loaders)
+
+  executor = ycsb.YCSBExecutor('hbase-10')
+
+  metadata = {'ycsb_client_vms': FLAGS.ycsb_client_vms,
+              'num_vms': FLAGS.num_vms,
```
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: maybe 'hbase_cluster_size": max(FLAGS.num_vms, 2) is more descriptive, unless you were just trying to capture the relevant flags

File: perfkitbenchmarker/data/hbase/__init__.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2014 Google Inc. All rights reserved.
```
Comment (ehankland) [inline perfkitbenchmarker/data/hbase/__init__.py]: 2015

File: perfkitbenchmarker/data/hbase/create-ycsb-table.hbaseshell.j2 [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2014 Google Inc. All rights reserved.
```
Comment (ehankland) [inline perfkitbenchmarker/data/hbase/create-ycsb-table.hbaseshell.j2]: 2015

File: perfkitbenchmarker/packages/ycsb.py [hunk 10]
```diff
@@ -12,31 +12,127 @@
...
+_DEFAULT_PERCENTILES = 50, 75, 90, 95, 99, 99.9
+
+
+flags.DEFINE_boolean('ycsb_histogram', True, 'Include individual '
+                     'histogram results from YCSB (will increase sample '
+                     'count).')
+flags.DEFINE_boolean('ycsb_include_individual_results', False,
+                     'Include results from each client VM, rather than just '
+                     'combined results.')
+flags.DEFINE_integer('ycsb_client_vms', 1, 'Number of YCSB client VMs.')
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: may want to include lower_bound=1

File: perfkitbenchmarker/packages/ycsb.py [hunk 11]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+        'statistics': {}
+    }
+    for _, name, val in lines:
+      name = name.strip()
+      val = val.strip()
+      # Drop ">" from ">1000"
+      if name.startswith('>'):
+        name = name[1:]
+      if name.isdigit():
+        val = float(val) if '.' in val else int(val)
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: why not pull this line up above if name.isdigit() and get rid of lines 227-230 below?

File: perfkitbenchmarker/packages/ycsb.py [hunk 12]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+  """Parse a YCSB workload file.
+
+  Args:
+    contents: str. Contents of the file.
+
+  Returns:
+    dict mapping from property key to property value for each property found in
+    'contents'.
+  """
+  matches = re.findall(r'^([^#]+?)\s*=\s*(.*)$', contents, re.MULTILINE)
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: is it possible to have k=v  #comment or are there only full line comments?

File: perfkitbenchmarker/packages/ycsb.py [hunk 13]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+    self.parameter_files = parameter_files or []
+    self.parameters = kwargs.copy()
+
+  def _BuildCommand(self, command_name, parameter_files=None, **kwargs):
+    command = [YCSB_EXE, command_name, self.database]
+
+    parameters = self.parameters.copy()
+    parameters.update(kwargs)
+
+    # these are passed as flags rather than properties, so they
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: start comment with uppercase

File: perfkitbenchmarker/packages/ycsb.py [hunk 14]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+    for param_file in list(self.parameter_files) + list(parameter_files or []):
+      command.extend(('-P', param_file))
+
+    for parameter, value in parameters.iteritems():
+      command.extend(('-p', '{0}={1}'.format(parameter, value)))
+
+    return ' '.join(command)
+
+  def _Load(self, vm, **kwargs):
+    """Execute ycsb load on 'vm'."""
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: capitalize YCSB

File: perfkitbenchmarker/packages/ycsb.py [hunk 15]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+    target = kwargs.pop('target', None)
+    if target is not None:
+      target_per_client = target // len(vms)
+      targets = [target_per_client +
+                 (1 if i < (target % len(vms)) else 0)
+                 for i in xrange(len(vms))]
+    else:
+      targets = [target for vm in vms]
+
+    result = []
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: plural?

File: perfkitbenchmarker/packages/ycsb.py [hunk 16]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+      workloads: List of strings. Workload files to use. If unspecified,
+        _GetWorkloadFileList() is used.
+      load_kwargs: dict. Additional arguments to pass to the load stage.
+      run_kwargs: dict. Additional arguments to pass to the run stage.
+    Returns:
+      List of sample.Sample objects.
+    """
+    workloads = workloads or _GetWorkloadFileList()
+    assert workloads, 'no workloads'
+    load_samples = list(self._LoadThreaded(vms, workloads[0],
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: Should we provide the option to skip reporting load_samples if people only care about the run samples?

File: perfkitbenchmarker/packages/ycsb.py [hunk 17]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+  def _RunThreaded(self, vms, **kwargs):
+    """Run a single workload using `vms`."""
+    target = kwargs.pop('target', None)
+    if target is not None:
+      target_per_client = target // len(vms)
+      targets = [target_per_client +
+                 (1 if i < (target % len(vms)) else 0)
+                 for i in xrange(len(vms))]
+    else:
+      targets = [target for vm in vms]
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: targets = [target for _ in vms] ?

File: perfkitbenchmarker/packages/ycsb.py [hunk 18]
```diff
@@ -47,3 +143,533 @@ def YumInstall(vm):
...
+      percentile = int(percentile)
+    label = 'p{0}'.format(percentile)
+    latencies, freqs = zip(*histogram)
+    time_ms = _WeightedQuantile(latencies, freqs, percentile * 0.01)
+    result[label] = time_ms
+  return result
+
+
+def _CombineResults(result_list, combine_histograms=True):
+  """Combine results from multiple YCSB clients."""
```
Comment (ehankland) [inline perfkitbenchmarker/packages/ycsb.py]: maybe include arg/returns description in docstring since it's a longer function

**Comment by ehankland** [comment]
Nice - some small comments.

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py (hunk 1 above)
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: Nope. Moved below.

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py (hunk 2 above)
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: Done.

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py (hunk 3 above)
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: Done.

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py (hunk 4 above)
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: Done for each assert.

File: perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py (hunk 7 above)
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/hbase_ycsb_benchmark.py]: Renamed: `hbase_cluster_size: len(by_role['hbase_vms'])`

File: perfkitbenchmarker/data/hbase/__init__.py (hunk 8 above)
Comment (cmccoy) [inline perfkitbenchmarker/data/hbase/__init__.py]: Done.

File: perfkitbenchmarker/data/hbase/create-ycsb-table.hbaseshell.j2 (hunk 9 above)
Comment (cmccoy) [inline perfkitbenchmarker/data/hbase/create-ycsb-table.hbaseshell.j2]: Done.

File: perfkitbenchmarker/packages/ycsb.py (hunk 10 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: Done - thanks.

File: perfkitbenchmarker/packages/ycsb.py (hunk 11 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: Good idea - done.

File: perfkitbenchmarker/packages/ycsb.py (hunk 18 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: Done.

File: perfkitbenchmarker/packages/ycsb.py (hunk 12 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: As best I can tell, [no](http://en.wikipedia.org/wiki/.properties). Agreed that this is fragile though. I've made it a little cleaner, and noted the limitations in the docstring.

File: perfkitbenchmarker/packages/ycsb.py (hunk 13 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: Done.

File: perfkitbenchmarker/packages/ycsb.py (hunk 14 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: This is actually meant to be a command (`ycsb load`). Added quotes.

File: perfkitbenchmarker/packages/ycsb.py (hunk 17 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: Done.

File: perfkitbenchmarker/packages/ycsb.py (hunk 15 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: done.

File: perfkitbenchmarker/packages/ycsb.py (hunk 16 above)
Comment (cmccoy) [inline perfkitbenchmarker/packages/ycsb.py]: Sure. Done.

**Comment by cmccoy** [comment]
@ehankland - thanks. Comments addressed in af1a91b5b709f81e13a8fa67988a408c69e83dc3. Running locally now to test. PTAL.

**Pull request merged**